src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs

```diff
@@ -273,80 +274,149 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
                 _tracer.Log($"Ovewriting existing File {bundlePath}");
             }
 
-            BinaryUtils.CopyFile(hostSource, bundlePath);
-
-            // Note: We're comparing file paths both on the OS we're running on as well as on the target OS for the app
-            // We can't really make assumptions about the file systems (even on Linux there can be case insensitive file systems
-            // and vice versa for Windows). So it's safer to do case sensitive comparison everywhere.
-            var relativePathToSpec = new Dictionary<string, FileSpec>(StringComparer.Ordinal);
-
             long headerOffset = 0;
-            using (FileStream bundle = File.Open(bundlePath, FileMode.Open, FileAccess.ReadWrite))
-            using (BinaryWriter writer = new BinaryWriter(bundle, Encoding.Default, leaveOpen: true))
+            string destinationDirectory = new FileInfo(bundlePath).Directory?.FullName ?? "";
+            if (!Directory.Exists(destinationDirectory))
             {
-                if (_target.IsOSX)
+                Directory.CreateDirectory(destinationDirectory);
+            }
+            using (FileStream bundleStream = File.Open(bundlePath, FileMode.Create, FileAccess.ReadWrite))
+            {
+                using (FileStream hostSourceStream = File.OpenRead(hostSource))
                 {
-                    MachObjectFile.RemoveCodeSignatureIfPresent(bundle);
+                    hostSourceStream.CopyTo(bundleStream);
                 }
-                bundle.Position = bundle.Length;
-                foreach (var fileSpec in fileSpecs)
+                EmbeddedSignatureBlob? signatureBlob = null;
+                if (_target.IsOSX)
                 {
-                    string relativePath = fileSpec.BundleRelativePath;
-
-                    if (IsHost(relativePath))
+                    long? newLength = null;
+                    using (MemoryMappedFile bundleMap = MemoryMappedFile.CreateFromFile(bundleStream, null, bundleStream.Length, MemoryMappedFileAccess.ReadWrite, HandleInheritability.None, leaveOpen: true))
```

Comment:
It feels weird to copy the file, then memory map the copied file.

If I remember correctly, the memory map API supports reading the input file, modifying the input, then writing the modifications to an output file.

Reply:
Ah, yeah we should be able to do that.

Reply:
Updated to create an in-memory memory-mapped file, write all the bundle data, then write out to the output path.